server/cache.go

```diff
@@ -101,32 +102,99 @@ func (s *storesInfo) getStoreCount() int {
 	return len(s.stores)
 }
 
+// regionMap wraps a map[uint64]*regionInfo and supports randomly pick a region.
+type regionMap struct {
+	m   map[uint64]*regionEntry
+	ids []uint64
+}
+
+type regionEntry struct {
+	*regionInfo
+	pos int
+}
+
+func newRegionMap() *regionMap {
+	return &regionMap{
+		m: make(map[uint64]*regionEntry),
+	}
+}
+
+func (m *regionMap) Len() int {
+	if m == nil {
```

Comment:
It's not for checking if inner map is empty. Consider the old code `delete(r.regions, region.GetId())` is changed to `r.regions.Delete(region.GetId())`, without this check, it need to be `if r.regions != nil { r.regions.Delete(region.GetId()) }`.